src/http-request.ts

```diff
@@ -31,3 +45,194 @@ export interface CloudLoggingHttpRequest {
+      req.headers[name] = value;
+    },
+    getHeader(name: string) {
+      return req.headers[name];
+    },
+  };
+  return wrapper;
+}
+
+export interface CloudTraceContext {
+  trace: string;
+  spanId?: string;
+  traceSampled?: boolean;
+}
+
+/**
+ * getTraceContext returns a CloudTraceContext with as many available trace and
+ * span properties as possible. It examines HTTP headers for trace context.
+ * Optionally, it can inject a W3C compliant trace context when no context is
+ * available from headers.
+ */
+export function getTraceContext(
+  req: http.IncomingMessage,
+  projectId: string,
+  inject?: boolean
+): CloudTraceContext {
+  const context: CloudTraceContext = {
+    trace: '',
+    spanId: undefined,
+    traceSampled: undefined,
+  };
+  const wrapper = makeHeaderWrapper(req);
+  if (wrapper) {
+    // Detect 'X-Cloud-Trace-Context' header.
+    const cloudContext = getCloudTraceContext(wrapper);
+    if (cloudContext) {
+      context.trace = `projects/${projectId}/traces/${cloudContext.trace}`;
+      context.spanId = cloudContext.spanId;
+      context.traceSampled = cloudContext.traceSampled;
+    } else {
+      // Detect 'traceparent' header, optionally injecting one if not available.
+      const parentContext = getOrInjectTraceParent(wrapper, inject);
+      if (parentContext) {
+        context.trace = `projects/${projectId}/traces/${parentContext.traceId}`;
+        context.spanId = parentContext.spanId;
+      }
+    }
+  }
+  return context;
+}
+
+/**
+ * getCloudTraceContext looks for the HTTP header 'X-Cloud-Trace-Context'
+ * per Cloud Trace specifications. `traceSampled` is false by default.
+ * @param headerWrapper
+ */
+export function getCloudTraceContext(
+  headerWrapper: HeaderWrapper
+): CloudTraceContext | null {
+  // Infer trace & span if not user specified already
+  const regex = /([a-f\d]+)?(\/?([a-f\d]+))?(;?o=(\d))?/;
+  const match = headerWrapper
+    .getHeader('x-cloud-trace-context')
+    ?.toString()
+    .match(regex);
+  if (match) {
+    return {
+      trace: match[1],
+      spanId: match[3],
+      traceSampled: match[5] === '1',
+    };
+  } else {
+    return null;
+  }
+}
+
+/**
+ * getCloudTraceContext looks for the HTTP header 'traceparent'
+ * per W3C specifications for OpenTelemetry and OpenCensus
+ * Read more about W3C protocol: https://www.w3.org/TR/trace-context/
+ *
+ * Optional: Inject a W3C compliant context with trace and span in the header,
+ * if not available.
+ *
+ * @param headerWrapper
+ * @param inject
+ */
+export function getOrInjectTraceParent(
+  headerWrapper: HeaderWrapper,
+  inject?: boolean
+): w3cContext.SpanContext | null {
+  let spanContext = w3cContext.extract(headerWrapper);
+  if (spanContext) return spanContext;
+  if (inject) {
+    // We were the first actor to detect lack of context. Establish context.
+    spanContext = w3cContext.generate();
+    w3cContext.inject(headerWrapper, spanContext);
```

Comment:
I like it but behavior might be not consistent:
If **x-cloud-trace-context** is present, we extract it and pass over.
If not present, with this logic, we'll create a new traceparent (different format for context propagation than what Cloud Trace supports natively).

So what can happen:
- A system has service A that gets incoming requests from external systems with trace info specified in Cloud Trace format. In addition, there is periodic job that triggers the middleware w/o providing the trace/span info.
- When the request comes with x-cloud-trace-context, it will be propagated downstream in format X
- When periodic job is triggered, the request has no trace/span, so it will be propagated with traceparent - a format Y != X. 

We probably need to make it configurable - the default trace format to use when creating a new context.

Reply:
Right now, we have to automatically inject `traceparent` if user doesn't have trace context && is using middleware. This is the default/previous behavior so turning it off would be a breaking change. 

In the non-middleware scenario, we actually don't inject a w3c context into the header (see `entry.ts`. The inject parameter is an optional bool, see:
```js
export function getOrInjectTraceParent(
  headerWrapper: HeaderWrapper,
  inject?: boolean
): w3cContext.SpanContext | null {
```